fcmp++: Consolidate paths in RPC to get path by unified ID

j-berman opened the pull request:
Serialization guru, the `ConsolidatedPaths` struct is using serializable type helpers that are strictly used for serializing/de-serializing, and I wonder if there may be a reasonably simpler approach there

I also hardened the `scan_tx` tests to also test for spending immediately after scanning.

j-berman committed:
fcmp++: Consolidate paths in RPC to get path by output ID

j-berman committed:
fcmp++: Consolidate paths in RPC to get path by output ID

jeffro256 commented:
Any metrics on how much this actually helps data size?

j-berman commented:
Let's say you're requesting paths for the first 38 outputs in the tree, the tree has 6 layers, and the tree is 100% full.

Individual path = 32 bytes\*(38\*3+18\*2+1) = 4832 bytes

**Without** consolidating, then the response is roughly 4832 bytes*38 = 183,616 bytes
**With** consolidating, then it's just 4832 bytes (since the first 38 outputs in the tree all share the same path)

So in the worst (best?) case, the response should be 38x smaller with this PR

j-berman commented:
Presumably people will end up requesting paths for outputs that are somewhat close in the tree (people are probably more likely to spend closer together than further apart time-wise), in which case there's decent benefit from this PR. Also this endpoint may be used for decoy selection algos at some point, which would bias toward recent (and with "binning", would select outputs right next to each other too). I figure makes sense for the RPC to be optimized now since it's pretty simple to do, and would be a little annoying to do later (would probably just implement a totally new endpoint and then we're maintaining both).

jeffro256 commented on src/blockchain_db/blockchain_db.h:
```diff
@@ -1884,7 +1884,7 @@ class BlockchainDB
   virtual uint64_t get_path_by_global_output_id(const std::vector<uint64_t> &global_output_ids,
     const uint64_t as_of_n_blocks,
-    std::vector<uint64_t> &leaf_idxs_out,
+    std::vector<fcmp_pp::curve_trees::AssignedLeafIdx> &leaf_idxs_out,
```
Sorry I'm slightly confused why this was changed? Aren't all of them going to be assigned or else the whole thing fails?

j-berman replied:
> Aren't all of them going to be assigned

If a user requests a coinbase output ID 15 blocks back from tip e.g., it won't be assigned

> or else the whole thing fails

If the client in `scan_tx` requests an output not yet in the tree, it fails client-side. I figure it makes sense for the daemon to still be able to return a success response in this case though, since the endpoint is expected to have wider usage beyond just `scan_tx`.

jeffro256 commented on src/fcmp_pp/curve_trees.h:
```diff
@@ -211,6 +211,108 @@ struct PathIndexes final
+        END_KV_SERIALIZE_MAP()
+    };
+
+    std::vector<LeafChunk> leaves_by_chunk_idx_vec;
```
Why don't these go inside `BEGIN_KV_SERIALIZE_MAP`?

j-berman replied:
Because the KV_SERIALIZE macros require them to be members of the parent struct as written :/

EDIT: moved in the latest, just called the function the macros are calling directly (I saw @vtnerd do this in p2p ssl)

jeffro256 commented on src/fcmp_pp/curve_trees.h:
```diff
@@ -211,6 +211,108 @@ struct PathIndexes final
+    std::vector<std::unordered_map<uint64_t, ChunkBytes>> layer_chunks_by_chunk_idx;
+
+    // Serializable helper types
+    struct LeafChunk
```
These types can be `private`

jeffro256 commented on src/fcmp_pp/curve_trees.h:
```diff
@@ -211,6 +211,108 @@ struct PathIndexes final
+        std::vector<OutputContext> chunk;
+
+        BEGIN_KV_SERIALIZE_MAP()
+            KV_SERIALIZE(chunk_idx)
```
These field names will be duplicated in the payload a lot, is that acceptable for a consolidated format?

j-berman replied:
This was a good spot. I de-duplicated in the latest, also de-duplicating the assigned leaf idxs.

Note: I can do the same/similar for `init_tree_sync_data_t` which may be worth it before v1 release

j-berman committed:
FCMP++/Carrot mega commit [4/20/26]

Co-authored-by: jeffro256 <jeffro256@tutanota.com>
Co-authored-by: Luke Parker <lukeparker5132@gmail.com>
Co-authored-by: Lee *!* Clagett <code@leeclagett.com>
Co-authored-by: SyntheticBird45 <someoneelse.is_on.github.rio7x@simplelogin.com>
Co-authored-by: jeetrex17 <jeetrex@tutamail.com>
Co-authored-by: UkoeHB <ukoehb@users.noreply.github.com>

jeffro256 committed:
simplewallet: fix payment ID false positive on Carrot receive

Builds on https://github.com/monero-project/monero/pull/10189

jeffro256 commented:
Needs rebase

jeffro256 commented:
I think that since this RPC endpoint is being touched, it would be a good time to bring up changing the request from taking the DB-style global indices to the amount-style global indices, since amount-style indices are already widely used in the ecosystem for wallet<->daemon RPC calls, most notably in `get_blocks.bin`, whereas the DB-style global indices aren't, AFAIK. As far as I can tell, the only place the DB-style ones are used is for this RPC call, and they are fetched by `wallet2` using an RPC endpoint which *also* serves amount-style indices. @vtnerd mentioned an issue with requesting paths in the LWS because of the indexing system. I'm trying to implement FCMP++ spend proofs, which requires fetching the paths as of a certain block for certain outputs, for which `wallet2` has the amount-style indices on the sender side, but not the DB-style indices.

j-berman commented:
I disagree for the same reason as [here](https://github.com/seraphis-migration/monero/pull/62#discussion_r2231718983). It makes more sense to move away from amount-style indices long term than to carry the debt forwards and harden it in a new RPC endpoint, since the amount-style indices are not actually ~~useful~~ optimal* for anything but legacy ring signatures. That was also my reasoning for using global indices in this endpoint.

On wallet side for spend proofs, the tree cache stores global indices for received outputs, so it is retrievable there.

j-berman commented:
My bad, the tree cache doesn't actually save global indices long term either (only leaf idxs). Regardless, I think the better solution is either for the wallet to keep the global index saved too at some point, or to make the extra trip to the daemon, or implement a new endpoint that can take amount output ID's and then use this endpoint's logic internally (EDIT: or an endpoint to get global output ID's from amount output ID's). I think using global indices (not tethered to amount) is the more future-proof endpoint that has a stronger claim at v1 / long term maintenance.

jeffro256 commented:
> On wallet side for spend proofs, the tree cache stores global indices for received outputs, so it is retrievable there.

Does the wallet not trigger the tree cache to drop outputs that have been spent?

> I think using global indices (not tethered to amount) is the more future-proof endpoint that has a stronger claim at v1 / long term maintenance.

> or an endpoint to get global output ID's from amount output ID's

I think that if we're going to really commit to this indexing scheme, then yes, there should be an RPC endpoint to convert index types.

jeffro256 commented:
The RPC endpoint to convert amount-style indices to DB-style indices should be very easy to implement with the current DB scheme, with a simple lookup in the `output_amounts` table and reading the `outkey::output_id` field.

j-berman commented:
> Does the wallet not trigger the tree cache to drop outputs that have been spent?

It doesn't. I don't think we need that for v1 but it would be nice to have

> I think that if we're going to really commit to this indexing scheme, then yes, there should be an RPC endpoint to convert index types

I don't think it's fully necessary, it holds on to an index that should be deprecated. But if you feel it is/it's desired, then it seems worth a separate PR imo. I think my argument is sound for aiming to move forward with global output ID's

jeffro256 commented:
> It doesn't. I don't think we need that for v1 but it would be nice to have

Well actually, maybe my exact use case (spend proofs) is a reason *against* dropping outputs after they're spent: it erases the privacy impact of requesting output paths from the daemon to create them. And the storage cost per output is only $O(log(N))$, where $N$ is the number of leaf tuples in the tree. There might be one problem though: a spend proof needs to get the historical path of output given its reference block number, so the tree cache would need to track the first reference block for each node where the node becomes available in the tree. Not only that, it would have to do that for each child of nodes we want to keep in the tree, which would otherwise have been pruned, so that it can recalculate nodes on the "edge" of the path at the desired reference block number. Gross...

It's easy for the daemon to calculate the path at any given reference block number because it knows A) all leaf tuples and B) the number of leaf tuples at each block. So maybe for v1 we don't even try to keep the historical path fetching local, and always rely on the daemon. To do that now, I can make the `/get_transactions` requests and use the returned `global_output_ids` in the response, then call `/get_path_by_global_output_id.bin` with those global IDs. I should take care to include all the global IDs in the TX, not just the ones we need for the proof for a sliver of privacy. This use case would definitely benefit from consolidated paths in the RPC.

navidR committed:
tx relay v2: rework tx request management and relay logic

    Implement new relay logic with threshold-based peer dropping, single-peer
    tx hash requests, per-connection in-flight limits with queuing, and
    multiindex-based request tracking. Consolidate add/remove request paths,
    pass tx hashes directly into send_txs/relay_txs, and add unit tests for
    the request manager.

    Co-authored-by: j-berman <justinberman@protonmail.com>
    Co-authored-by: Boog900 <boog900@tutanota.com>

vtnerd committed:
Fix ZMQ-PUB reporting of mempool txes that were initially in stem phase

vtnerd commented:
#278 has a ZMQ endpoint for converting legacy output ids to unified ids. It also switches ZMQ get_blocks to unified indices after the hard-fork. The "catch" is whether wallets should immediately switchover after the hard-fork, or whether they should do it lazily when needed.

I'm in the process of rebasing + updating that PR btw.

vtnerd commented:
Also, if you use that ZMQ conversion endpoint, it effectively leaks your outputs to the daemon. I'm not sure how to do the conversion to unified in a privacy preserving way - I guess the wallet could ask for all values or something.

j-berman committed:
Benchmark FCMP++ verification (using lib w/latest alpha optimizations)

j-berman committed:
Benchmark FCMP++ with latest fix to GBP's applied

j-berman committed:
Fix max tx weight limit for FCMP++ txs

Fixes https://github.com/seraphis-migration/monero/issues/314

j-berman committed:
Bin pack FCMP++ batch verification

j-berman committed:
More tests for benchmarking FCMP++ verification

j-berman committed:
Use older versions of crates so rust 1.69.0 works

j-berman committed:
Use upstream monero-oxide lib

jeffro256 commented on src/fcmp_pp/tree_cache.h:
```diff
@@ -103,20 +103,6 @@ struct CachedTreeElemChunk final
     END_SERIALIZE()
 };
 
-struct AssignedLeafIdx final
```
Why were these types hoisted into `fcmp_pp_types.h`? And can they be moved into `curve_trees.h` or `tree_cache.h`?

j-berman replied:
It's now also used in the RPC layer, which doesn't require a caller use `curve_trees.h` or `tree_cache.h` (similar logic for `CompressedTreeExtension` used [here](https://github.com/seraphis-migration/monero/pull/286#discussion_r2856674363)).

jeffro256 committed:
carrot_core: test ECDH convergence with OpenSSL

j-berman committed:
tx_pool: notify txpool event when stem bumps to fluff

j-berman committed:
Bump FCMP++ reference tx weight as a result of larger proofs

https://libera.monerologs.net/monero-research-lab/20260415#c667853-c667866

tobtoht committed:
ci: MSRV 1.69 -> 1.93

j-berman committed:
@jeffro256 review comments

j-berman committed:
Rm stray filesystem addition

j-berman committed:
n_tree_layers is safe as a uint8_t, varint not necessary

j-berman committed:
Fix rebase

j-berman committed:
Revert "Use older versions of crates so rust 1.69.0 works"

This reverts commit e7fb6ab2c46795c3fbb75f2d243dfd52b461f860.

j-berman committed:
@jeffro256 comment

jeffro256 committed:
Merge pull request #319 from j-berman/use-latest-fcmp++

Use latest FCMP++ lib changes

j-berman committed:
Merge pull request #323 from j-berman/n-layers-u8

n_tree_layers is safe as a uint8_t, varint not necessary

jeffro256 committed:
Merge pull request #321 from j-berman/bump-tx-ref-weight

Bump FCMP++ reference tx weight as a result of larger proofs

jeffro256 committed:
Merge pull request #320 from j-berman/max-tx-weight

Include FCMP_PLUS_PLUS_MAX_TX_WEIGHT

jeffro256 committed:
beta: fix functional tests base fee estimate

jeffro256 committed:
Merge pull request #326 from jeffro256/fcmp++-zmq-pub-fixes

Fcmp++ zmq pub fixes

jeffro256 committed:
Merge pull request #217 from jeffro256/cli_wallet_pid_recv_warn

simplewallet: fix payment ID false positive on Carrot receive

jeffro256 committed:
tests: fix linking issue for `unit_tests::data_dir` variable

jeffro256 committed:
carrot: remove trailing whitespace

jeffro256 committed:
Merge pull request #327 from jeffro256/carrot_core_whitespace

carrot: remove trailing whitespace

jeffro256 committed:
wallet: add fork checks to wallet2::transfer_selected{_rct}

https://github.com/seraphis-migration/monero/pull/52#issuecomment-4291832590

j-berman committed:
Merge pull request #325 from jeffro256/beta_func_test_fee_estim

beta: fix functional tests base fee estimate

jeffro256 committed:
Merge pull request #328 from jeffro256/fcmp++-transfer-selected-check

wallet: add fork checks to `wallet2::transfer_selected{_rct}`

jeffro256 committed:
Merge pull request #318 from jeffro256/carrot_core_ssl_x25519

carrot_core: test ECDH convergence with OpenSSL

jeffro256 committed:
blockchain_utilities: fix --data-dir option

tobtoht committed:
trezor: replace use of boost::is_base_of with std::is_base_of_v

j-berman committed:
Use latest round of changes to FCMP++ lib

jeffro256 committed:
Merge pull request #330 from j-berman/latest-lib-rd2

Use latest round of changes to FCMP++ lib

jeffro256 committed:
Merge pull request #329 from jeffro256/fix_bc_util_args_fcmp++

blockchain_utilities: fix `--data-dir` option [FCMP++]

j-berman committed:
Update FCMP++ max tx weight & FCMP++ proof bins

j-berman committed:
Merge pull request #332 from j-berman/update-for-latest-fcmp++

Update FCMP++ max tx weight & FCMP++ proof bins

jeffro256 committed:
carrot_impl: fix addr derive type log spam

Co-authored-by: jeetrex17 <jeetrex@tutamail.com>

jeffro256 committed:
Merge pull request #336 from jeffro256/expected_derive_type_log

carrot_impl: fix addr derive type log spam
